services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,528 @@
+                        tenantDtoResult.getTenant(),
+                        Optional.ofNullable(DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())),
+                        Optional.ofNullable(tenantDtoResult.getVersion()))));
+
+    }
+
+    /**
+     * Gets a tenant dto by tenant id.
+     *
+     * @param tenantId the tenant id.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future with a {@link TenantDto} be provided if found.
+     *         Otherwise the future will fail with
+     *         a {@link ClientErrorException} with {@link HttpURLConnection#HTTP_NOT_FOUND}.
+     * @throws NullPointerException if the parameter is {@code null}.
+     */
+    private Future<TenantDto> findTenant(final String tenantId) {
+        Objects.requireNonNull(tenantId);
+
+        final JsonObject findTenantQuery = new MongoDbDocumentBuilder()
+                .withTenantId(tenantId)
+                .document();
+        return findTenant(findTenantQuery);
+    }
+
+    /**
+     * Fetches tenant by subject DN.
+     *
+     * @param subjectDn the subject DN.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future with a {@link TenantDto} be provided if found.
+     *         Otherwise the future will fail with
+     *         a {@link ClientErrorException} with {@link HttpURLConnection#HTTP_NOT_FOUND}.
+     * @throws NullPointerException if the parameter is {@code null}.
+     */
+    private Future<TenantDto> findTenant(final X500Principal subjectDn) {
+        Objects.requireNonNull(subjectDn);
+
+        final JsonObject findTenantQuery = new MongoDbDocumentBuilder()
+                .withCa("tenant", subjectDn.getName())
+                .document();
+
+        return findTenant(findTenantQuery);
+    }
+
+    /**
+     * Gets a tenant dto by tenant id.
+     *
+     * @param findQuery the tenant query.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future with a {@link TenantDto} be provided if found.
+     *         Otherwise the future will fail with
+     *         a {@link ClientErrorException} with {@link HttpURLConnection#HTTP_NOT_FOUND}.
+     * @throws NullPointerException if the parameter is {@code null}.
+     */
+    private Future<TenantDto> findTenant(final JsonObject findQuery) {
+        Objects.requireNonNull(findQuery);
+
+        final Promise<JsonObject> findTenantPromise = Promise.promise();
+        mongoClient.findOne(config.getCollectionName(), findQuery, new JsonObject(), findTenantPromise);
+        return findTenantPromise.future()
+                .compose(tenantJsonResult -> Optional.ofNullable(tenantJsonResult)
+                        .map(tenantJson -> Future.succeededFuture(tenantJson.mapTo(TenantDto.class)))
+                        .orElse(Future.failedFuture(new NullPointerException()))
```

Comment:
Why this `NullPointerException` instead directly a `ClientErrorException` could be returned as in the `MongoDbBasedRegistrationService`?

Also I didn't get why it is being checked if `findQuery` contains `FIELD_PAYLOAD_TENANT_ID` and then same HTTP_NOT_FOUND status is returned even if it is available or not.

Reply:
The check was necessary due to optional tenantId.
Function is simplified now.